lib/puppet/file_system.rb

```diff
@@ -122,8 +122,8 @@ def self.each_line(path, &block)
   #
   # @api public
   #
-  def self.read(path)
-    @impl.read(assert_path(path))
+  def self.read(path, opts = {})
+    @impl.read(assert_path(path), opts)
```

Comment:
I should also note that I spiked a changeover of the remaining `File.read` calls in a separate branch last week at https://github.com/Iristyle/puppet/commits/ticket/stable/PUP-5728-convert-remaining-File-read-calls-to-Puppet-FileSystem .  I figured these calls weren't crucial to the ticket at hand, but should also be changed as part of another cleanup effort later - that might be the better spot to introduce the warning you're suggesting?

I think I'd also prefer not to add a new `File.read_characters` method - only because it's yet another special Puppet FileSystem behavior that implementers have to remember to use.